File: src/core/consolidate/identity.ts
import type { Finding } from "./findings.js";

/**
 * A stable name for a finding, and a fingerprint of what it currently says.
 *
 * Marking a finding "solved" has to survive a re-read of the board, and the
 * board is re-read every time you save. The *id* is what the finding is about
 * — this value, this key, these parts — and does not move when a count
 * changes. The *signature* is what the finding says right now.
 *
 * A mark is honoured only while both match. So settling "10k in 0402 and 0603"
 * keeps quiet through unrelated edits, and speaks up again the day a third
 * package appears: you settled the question that was asked, not the subject.
 */
export function findingId(f: Finding): string {
  switch (f.kind) {
    case "duplicate-spelling":
    case "mixed-footprint":
    case "singleton":
      return `${f.kind}:${f.key}`;
    case "multi-package":
      return `${f.kind}:${f.cls}:${f.value}${f.unit}`;
    case "near-value":
      return `${f.kind}:${f.members
        .map((m) => m.key)
        .sort()
        .join(",")}`;
  }
}

export function findingSignature(f: Finding): string {
  switch (f.kind) {
    case "duplicate-spelling":
      return f.spellings
        .map((s) => `${s.value}×${s.placements}`)
        .sort()
        .join(" ");
    case "mixed-footprint":
      return f.footprints
        .map((x) => `${x.footprint}×${x.placements}`)
        .sort()
        .join(" ");
    case "multi-package":
      return f.groups
        .map((g) => `${g.pkg}×${g.placements}`)
        .sort()
        .join(" ");
    case "near-value":
      return f.members
        .map((m) => `${m.value}×${m.placements}`)
        .sort()
        .join(" ");
    case "singleton":
      return f.ref;
  }
}

/** What the UI shows on a settled card, and what the file records. */
export type SolvedMark = {
  id: string;
  /** The finding's wording when it was settled; a change brings it back. */
  signature: string;
  at: string;
  note?: string;
};

File: kinv/core/consolidate/identity.py
"""A stable name for a finding, and a fingerprint of what it currently says.

Marking a finding "solved" has to survive a re-read of the board, and the
board is re-read every time you save. The *id* is what the finding is about —
this value, this key, these parts — and does not move when a count changes.
The *signature* is what the finding says right now.

A mark is honoured only while both match. So settling "10k in 0402 and 0603"
keeps quiet through unrelated edits, and speaks up again the day a third
package appears: you settled the question that was asked, not the subject.
"""

from __future__ import annotations

from kinv.js import code_units, num_str


def _joined(items: list[str], sep: str) -> str:
    return sep.join(sorted(items, key=code_units))


def finding_id(f: dict) -> str:
    kind = f["kind"]
    if kind in ("duplicate-spelling", "mixed-footprint", "singleton"):
        return f"{kind}:{f['key']}"
    if kind == "multi-package":
        return f"{kind}:{f['cls']}:{f['value']}{f['unit']}"
    return f"{kind}:{_joined([m['key'] for m in f['members']], ',')}"


def finding_signature(f: dict) -> str:
    kind = f["kind"]
    times = "×"
    if kind == "duplicate-spelling":
        return _joined([f"{s['value']}{times}{num_str(s['placements'])}" for s in f["spellings"]], " ")
    if kind == "mixed-footprint":
        return _joined([f"{x['footprint']}{times}{num_str(x['placements'])}" for x in f["footprints"]], " ")
    if kind == "multi-package":
        return _joined([f"{g['pkg']}{times}{num_str(g['placements'])}" for g in f["groups"]], " ")
    if kind == "near-value":
        return _joined([f"{m['value']}{times}{num_str(m['placements'])}" for m in f["members"]], " ")
    return f["ref"]
